src/pymmcore_plus/mda/metadata/schema.py

```diff
@@ -0,0 +1,169 @@
+from typing import Any, Dict, List, Literal, Optional, Tuple, TypedDict
+
+import useq
+from typing_extensions import NotRequired
+
+__all__ = [
+    "FrameMetaV1",
+    "SummaryMetaV1",
+    "ConfigGroup",
+    "ConfigPreset",
+    "DeviceInfo",
+    "ImageInfo",
+    "PixelSizeConfigPreset",
+    "Position",
+    "PropertyInfo",
+    "PropertyValue",
+    "SystemInfo",
+]
+
+
+class PropertyInfo(TypedDict):
+    """Information about a single device property."""
+
+    name: str
+    value: Optional[str]
+    data_type: Literal["undefined", "float", "int", "str"]
+    allowed_values: Optional[Tuple[str, ...]]
+    is_read_only: bool
+    is_pre_init: bool
+    limits: NotRequired[Tuple[float, float]]
+    sequenceable: bool
+    sequence_max_length: NotRequired[int]
+    # device_label: str
+
+
+class DeviceInfo(TypedDict):
+    """Information about a specific device."""
+
+    label: str
+    library: str
+    name: str
+    type: str
+    description: str
+    properties: Tuple[PropertyInfo, ...]
+    parent_label: Optional[str]  # None on hub devices
+
+    # state device only
+    labels: NotRequired[Tuple[str, ...]]
+    # focus device only
+    focus_direction: NotRequired[Literal["Unknown", "TowardSample", "AwayFromSample"]]
+    # hub device only
+    child_names: NotRequired[Tuple[str, ...]]
+    # stage or XYstage device only
+    is_stage_sequenceable: NotRequired[bool]
+    # stage device only
+    is_continuous_focus_drive: NotRequired[bool]
+    is_stage_linear_sequenceable: NotRequired[bool]
+    # camera device only
+    is_exposure_sequenceable: NotRequired[bool]
+
+
+class SystemInfo(TypedDict):
+    """General system information."""
+
+    pymmcore_version: str
+    pymmcore_plus_version: str
+    mmcore_version: str
+    device_api_version: str
+    device_adapter_search_paths: Tuple[str, ...]
+    system_configuration: Optional[str]
+    primary_log_file: str
+    circular_buffer_memory_footprint: int
+    buffer_total_capacity: int
+    buffer_free_capacity: int
+    timeout_ms: int
+    # remaining_image_count: int
+    continuous_focus_enabled: bool
+    continuous_focus_locked: bool
+    auto_shutter: bool
+
+
+class ImageInfo(TypedDict):
+    """Information about the current image structure."""
+
+    bytes_per_pixel: int
+    current_pixel_size_config: str
+    exposure: float
+    image_bit_depth: int
+    image_buffer_size: int
+    image_height: int
+    image_width: int
+    magnification_factor: float
+    number_of_camera_channels: int
+    number_of_components: int
+    pixel_size_affine: Tuple[float, float, float, float, float, float]
+    pixel_size_um: float
+    roi: List[int]
+    camera_device: str
+    multi_roi: NotRequired[Tuple[List[int], List[int], List[int], List[int]]]
+
+
+class Position(TypedDict):
+    """Represents a position in 3D space and focus."""
+
+    x: Optional[float]
+    y: Optional[float]
+    focus: Optional[float]
```

Comment:
The MMStudio MDA supports multiple XY stages and multiple Z stages. I suspect it will make sense to be able to record multiple stages' positions, even if there are primary XYZ coordinates that can be derived from them? The common case is motorized XYZ + piezo Z (+ piezo XY).

Also, calling the Z stage "focus" in position contexts is not always correct, because any single-axis stage (sometimes even rotary stages) could be used. Admittedly there may not be a correct answer here.

Reply:
a more generic model could be:


```python
class StagePosition(TypedDict):
    device_label: str
    position: float | tuple[float, ...] 

positions: list[StagePosition]
```

or (some variant like that)